{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "73ed3589",
   "metadata": {},
   "source": [
    "# Pre-download frequently used genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c32c99",
   "metadata": {},
   "source": [
    "Download genomes from the orthology table and convert them to a sorted BED file.\n",
    "\n",
    "**Input**: Orthology table (.xsl, .csv, .tsv), version of Ensembl compara used, path to directory to save genome files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "26641730",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pathlib\n",
    "import urllib.request\n",
    "import gget\n",
    "import requests\n",
    "import sys\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2f30e45f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter the deisred ensembl release\n",
    "ensembl_release = 109"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7f6124e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter path to genomes directory\n",
    "downloaded_genomes_dir = pathlib.Path(\"genomes/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2f37ae91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter path to orthology table\n",
    "orthology_table_file = pathlib.Path(\"/home/vecerkok/orthology_table/orthology_table-68species_ensembl_ids.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2552b524",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthology_table(table_file: str):\n",
    "    table_df = pd.read_csv(table_file, low_memory=False, index_col=0)\n",
    "    species_names = table_df.columns\n",
    "    \n",
    "    if file_type == \".csv\":\n",
    "        return pd.read_csv(table_file, sep=\",\", header=0, index_col=0, low_memory=False)\n",
    "    elif file_type == \".tsv\":\n",
    "        return pd.read_csv(table_file, sep=\"\\t\", header=0, index_col=0, low_memory=False)\n",
    "    elif (file_type == \".xsls\") or (file_type == \".xsl\"):\n",
    "        return pd.read_excel(table_file, header=0, index_col=0, low_memory=False)\n",
    "    else:\n",
    "        print(\"Accepted file formats: .csv, .tsv, .xsls, .xsl!\")\n",
    "        return\n",
    "    \n",
    "    # modify species names to gtf filename format\n",
    "    species_names = [x.lower().replace(\" \", \"_\") for x in species_names]\n",
    "\n",
    "    species_names = list(map(lambda x: x.replace(\"heterocephalus_glaber\", \"heterocephalus_glaber_female\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"gorilla_gorilla_gorilla\", \"gorilla_gorilla\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"cricetulus_griseus\", \"cricetulus_griseus_chok1gshd\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"ovis_aries\", \"ovis_aries_rambouillet\"), species_names))\n",
    "    \n",
    "    table_df.columns = species_names\n",
    "    return table_df, species_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c9fff909",
   "metadata": {},
   "outputs": [],
   "source": [
    "orthology_table_df, species_names = get_orthology_table(orthology_table_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25d933b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun Apr 23 18:23:20 2023 INFO Fetching available genomes (GTF and FASTAs present) from Ensembl release 109.\n"
     ]
    }
   ],
   "source": [
    "# check if assemblies in orthology table correspond to assemblies in Ensembl db\n",
    "species_in_db_release = gget.ref(species=None, list_species=True, release=109) # list all species in release\n",
    "species_names = [x for x in species_names if x in species_in_db_release] # intersect of the two lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "831069f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_genome_annotation_downloaded(organism_name: str):\n",
    "    organism_genome_annotation = downloaded_genomes_dir / (organism_name + \".gtf\")\n",
    "    return organism_genome_annotation.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b26768d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_genome_annotation(organism_name: str):\n",
    "    # get genome annotation gtf\n",
    "    gtf_ftp = gget.ref(organism_name, which=[\"gtf\"], release=ensembl_release)[organism_name][\"annotation_gtf\"][\"ftp\"]\n",
    "    print(gtf_ftp)\n",
    "    \n",
    "    # download gtf genome nanotation\n",
    "    genome_file = downloaded_genomes_dir / (organism_name + \".gtf.gz\")\n",
    "    urllib.request.urlretrieve(gtf_ftp, genome_file)\n",
    "    \n",
    "    # decompress the genome annotation\n",
    "    !gzip -d $genome_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26e84302",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_annotation(organism_name: str):\n",
    "    if is_genome_annotation_downloaded(organism_name):\n",
    "        print(\"Genome annotation already downloaded and decompressed!\")\n",
    "    else:\n",
    "        print(\"Downloading and then decompressing the genome annotation!\")\n",
    "        download_genome_annotation(organism_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9822de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "homo_sapiens\n",
      "Genome annotation already downloaded and decompressed!\n",
      "cricetulus_griseus_chok1gshd\n",
      "Genome annotation already downloaded and decompressed!\n",
      "mesocricetus_auratus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "rattus_norvegicus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "cavia_porcellus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "octodon_degus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "heterocephalus_glaber_female\n",
      "Genome annotation already downloaded and decompressed!\n",
      "maylandia_zebra\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pteropus_vampyrus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "monodelphis_domestica\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pelodiscus_sinensis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "anolis_carolinensis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pundamilia_nyererei\n",
      "Genome annotation already downloaded and decompressed!\n",
      "vicugna_pacos\n",
      "Genome annotation already downloaded and decompressed!\n",
      "otolemur_garnettii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "takifugu_rubripes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "neolamprologus_brichardi\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chinchilla_lanigera\n",
      "Genome annotation already downloaded and decompressed!\n",
      "sorex_araneus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ictidomys_tridecemlineatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "geospiza_fortis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "jaculus_jaculus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "myotis_lucifugus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "taeniopygia_guttata\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ficedula_albicollis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chlorocebus_sabaeus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "nomascus_leucogenys\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gasterosteus_aculeatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "petromyzon_marinus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "latimeria_chalumnae\n",
      "Genome annotation already downloaded and decompressed!\n",
      "lepisosteus_oculatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "danio_rerio\n",
      "Genome annotation already downloaded and decompressed!\n",
      "microtus_ochrogaster\n",
      "Genome annotation already downloaded and decompressed!\n",
      "astyanax_mexicanus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gadus_morhua\n",
      "Genome annotation already downloaded and decompressed!\n",
      "xiphophorus_maculatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oryzias_latipes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oreochromis_niloticus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "haplochromis_burtoni\n",
      "Genome annotation already downloaded and decompressed!\n",
      "xenopus_tropicalis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chrysemys_picta_bellii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gallus_gallus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "meleagris_gallopavo\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ornithorhynchus_anatinus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "sarcophilus_harrisii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "notamacropus_eugenii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "dasypus_novemcinctus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "erinaceus_europaeus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "echinops_telfairi\n",
      "Genome annotation already downloaded and decompressed!\n",
      "callithrix_jacchus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "macaca_fascicularis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "macaca_mulatta\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gorilla_gorilla\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pan_troglodytes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pongo_abelii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "canis_lupus_familiaris\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ailuropoda_melanoleuca\n",
      "Genome annotation already downloaded and decompressed!\n",
      "mustela_putorius_furo\n",
      "Genome annotation already downloaded and decompressed!\n",
      "felis_catus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "tursiops_truncatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "loxodonta_africana\n",
      "Genome annotation already downloaded and decompressed!\n",
      "equus_caballus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "bos_taurus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "capra_hircus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ovis_aries_rambouillet\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ochotona_princeps\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oryctolagus_cuniculus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "tetraodon_nigroviridis\n",
      "Genome annotation already downloaded and decompressed!\n"
     ]
    }
   ],
   "source": [
    "# download assemblies one by one\n",
    "for species in species_names:\n",
    "    print(species)\n",
    "    get_genome_annotation(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c23aad0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gtf_to_bed(assembly_name):\n",
    "    # convert gtf file to bed format\n",
    "    gtf_file = downloaded_genomes_dir / (assembly_name + \".gtf\")\n",
    "    bed_file = downloaded_genomes_dir / (\"sorted_\" + assembly_name + \".bed\")\n",
    "    !gtf2bed < $gtf_file > $bed_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "83e394b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "homo_sapiens\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "cricetulus_griseus_chok1gshd\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "mesocricetus_auratus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "rattus_norvegicus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "cavia_porcellus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "octodon_degus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "heterocephalus_glaber_female\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "maylandia_zebra\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pteropus_vampyrus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "monodelphis_domestica\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pelodiscus_sinensis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "anolis_carolinensis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pundamilia_nyererei\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "vicugna_pacos\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "otolemur_garnettii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "takifugu_rubripes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "neolamprologus_brichardi\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "chinchilla_lanigera\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "sorex_araneus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ictidomys_tridecemlineatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "geospiza_fortis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "jaculus_jaculus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "myotis_lucifugus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "taeniopygia_guttata\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ficedula_albicollis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chlorocebus_sabaeus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "nomascus_leucogenys\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gasterosteus_aculeatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "petromyzon_marinus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "latimeria_chalumnae\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "lepisosteus_oculatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "danio_rerio\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "microtus_ochrogaster\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "astyanax_mexicanus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gadus_morhua\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "xiphophorus_maculatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oryzias_latipes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oreochromis_niloticus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "haplochromis_burtoni\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "xenopus_tropicalis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "chrysemys_picta_bellii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gallus_gallus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "meleagris_gallopavo\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ornithorhynchus_anatinus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "sarcophilus_harrisii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "notamacropus_eugenii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "dasypus_novemcinctus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "erinaceus_europaeus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "echinops_telfairi\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "callithrix_jacchus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "macaca_fascicularis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "macaca_mulatta\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gorilla_gorilla\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pan_troglodytes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pongo_abelii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "canis_lupus_familiaris\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ailuropoda_melanoleuca\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "mustela_putorius_furo\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "felis_catus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "tursiops_truncatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "loxodonta_africana\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "equus_caballus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "bos_taurus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "capra_hircus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ovis_aries_rambouillet\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ochotona_princeps\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oryctolagus_cuniculus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "tetraodon_nigroviridis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    }
   ],
   "source": [
    "# generate sorted bed files from genome annotations\n",
    "for species in species_names:\n",
    "    print(species)\n",
    "    gtf_to_bed(species)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
