{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "73ed3589",
   "metadata": {},
   "source": [
    "# Pre-download frequently used genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "26641730",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pathlib\n",
    "import urllib.request\n",
    "import gget\n",
    "import requests\n",
    "import sys\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2f30e45f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_release = 109"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a7f6124e",
   "metadata": {},
   "outputs": [],
   "source": [
    "downloaded_genomes_dir = pathlib.Path(\"genomes/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2f37ae91",
   "metadata": {},
   "outputs": [],
   "source": [
    "orthology_table_file = pathlib.Path(\"/home/vecerkok/orthology_table/orthology_table-68species_ensembl_ids.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2552b524",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthology_table(table_path):\n",
    "    orthology_table_df = pd.read_csv(orthology_table_file, low_memory=False).drop(columns=\"Unnamed: 0\")\n",
    "    species_names = orthology_table_df.columns\n",
    "    \n",
    "    # modify names to gtf filename format\n",
    "    species_names = [x.lower().replace(\" \", \"_\") for x in species_names]\n",
    "\n",
    "    species_names = list(map(lambda x: x.replace(\"heterocephalus_glaber\", \"heterocephalus_glaber_female\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"gorilla_gorilla_gorilla\", \"gorilla_gorilla\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"cricetulus_griseus\", \"cricetulus_griseus_chok1gshd\"), species_names))\n",
    "    species_names = list(map(lambda x: x.replace(\"ovis_aries\", \"ovis_aries_rambouillet\"), species_names))\n",
    "    \n",
    "    orthology_table_df.columns = species_names\n",
    "    return orthology_table_df, species_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c9fff909",
   "metadata": {},
   "outputs": [],
   "source": [
    "orthology_table_df, species_names = get_orthology_table(orthology_table_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "25d933b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mon Feb 20 20:52:10 2023 INFO Fetching available genomes (GTF and FASTAs present) from Ensembl release 109.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['acanthochromis_polyacanthus',\n",
       " 'accipiter_nisus',\n",
       " 'ailuropoda_melanoleuca',\n",
       " 'amazona_collaria',\n",
       " 'amphilophus_citrinellus',\n",
       " 'amphiprion_ocellaris',\n",
       " 'amphiprion_percula',\n",
       " 'anabas_testudineus',\n",
       " 'anas_platyrhynchos',\n",
       " 'anas_platyrhynchos_platyrhynchos',\n",
       " 'anas_zonorhyncha',\n",
       " 'anolis_carolinensis',\n",
       " 'anser_brachyrhynchus',\n",
       " 'anser_cygnoides',\n",
       " 'aotus_nancymaae',\n",
       " 'apteryx_haastii',\n",
       " 'apteryx_owenii',\n",
       " 'apteryx_rowi',\n",
       " 'aquila_chrysaetos_chrysaetos',\n",
       " 'astatotilapia_calliptera',\n",
       " 'astyanax_mexicanus',\n",
       " 'astyanax_mexicanus_pachon',\n",
       " 'athene_cunicularia',\n",
       " 'balaenoptera_musculus',\n",
       " 'betta_splendens',\n",
       " 'bison_bison_bison',\n",
       " 'bos_grunniens',\n",
       " 'bos_indicus_hybrid',\n",
       " 'bos_mutus',\n",
       " 'bos_taurus',\n",
       " 'bos_taurus_hybrid',\n",
       " 'bubo_bubo',\n",
       " 'buteo_japonicus',\n",
       " 'caenorhabditis_elegans',\n",
       " 'cairina_moschata_domestica',\n",
       " 'calidris_pugnax',\n",
       " 'calidris_pygmaea',\n",
       " 'callithrix_jacchus',\n",
       " 'callorhinchus_milii',\n",
       " 'camarhynchus_parvulus',\n",
       " 'camelus_dromedarius',\n",
       " 'canis_lupus_dingo',\n",
       " 'canis_lupus_familiaris',\n",
       " 'canis_lupus_familiarisbasenji',\n",
       " 'canis_lupus_familiarisboxer',\n",
       " 'canis_lupus_familiarisgreatdane',\n",
       " 'canis_lupus_familiarisgsd',\n",
       " 'capra_hircus',\n",
       " 'capra_hircus_blackbengal',\n",
       " 'carassius_auratus',\n",
       " 'carlito_syrichta',\n",
       " 'castor_canadensis',\n",
       " 'catagonus_wagneri',\n",
       " 'catharus_ustulatus',\n",
       " 'cavia_aperea',\n",
       " 'cavia_porcellus',\n",
       " 'cebus_imitator',\n",
       " 'cercocebus_atys',\n",
       " 'cervus_hanglu_yarkandensis',\n",
       " 'chelonoidis_abingdonii',\n",
       " 'chelydra_serpentina',\n",
       " 'chinchilla_lanigera',\n",
       " 'chlorocebus_sabaeus',\n",
       " 'choloepus_hoffmanni',\n",
       " 'chrysemys_picta_bellii',\n",
       " 'chrysolophus_pictus',\n",
       " 'ciona_intestinalis',\n",
       " 'ciona_savignyi',\n",
       " 'clupea_harengus',\n",
       " 'colobus_angolensis_palliatus',\n",
       " 'corvus_moneduloides',\n",
       " 'cottoperca_gobio',\n",
       " 'coturnix_japonica',\n",
       " 'cricetulus_griseus_chok1gshd',\n",
       " 'cricetulus_griseus_crigri',\n",
       " 'cricetulus_griseus_picr',\n",
       " 'crocodylus_porosus',\n",
       " 'cyanistes_caeruleus',\n",
       " 'cyclopterus_lumpus',\n",
       " 'cynoglossus_semilaevis',\n",
       " 'cyprinodon_variegatus',\n",
       " 'cyprinus_carpio_carpio',\n",
       " 'cyprinus_carpio_germanmirror',\n",
       " 'cyprinus_carpio_hebaored',\n",
       " 'cyprinus_carpio_huanghe',\n",
       " 'danio_rerio',\n",
       " 'dasypus_novemcinctus',\n",
       " 'delphinapterus_leucas',\n",
       " 'denticeps_clupeoides',\n",
       " 'dicentrarchus_labrax',\n",
       " 'dipodomys_ordii',\n",
       " 'dromaius_novaehollandiae',\n",
       " 'drosophila_melanogaster',\n",
       " 'echeneis_naucrates',\n",
       " 'echinops_telfairi',\n",
       " 'electrophorus_electricus',\n",
       " 'eptatretus_burgeri',\n",
       " 'equus_asinus',\n",
       " 'equus_asinus_asinus',\n",
       " 'equus_caballus',\n",
       " 'erinaceus_europaeus',\n",
       " 'erpetoichthys_calabaricus',\n",
       " 'erythrura_gouldiae',\n",
       " 'esox_lucius',\n",
       " 'falco_tinnunculus',\n",
       " 'felis_catus',\n",
       " 'ficedula_albicollis',\n",
       " 'fukomys_damarensis',\n",
       " 'fundulus_heteroclitus',\n",
       " 'gadus_morhua',\n",
       " 'gallus_gallus',\n",
       " 'gallus_gallus_gca000002315v5',\n",
       " 'gallus_gallus_gca016700215v2',\n",
       " 'gambusia_affinis',\n",
       " 'gasterosteus_aculeatus',\n",
       " 'geospiza_fortis',\n",
       " 'gopherus_agassizii',\n",
       " 'gopherus_evgoodei',\n",
       " 'gorilla_gorilla',\n",
       " 'gouania_willdenowi',\n",
       " 'haplochromis_burtoni',\n",
       " 'heterocephalus_glaber_female',\n",
       " 'heterocephalus_glaber_male',\n",
       " 'hippocampus_comes',\n",
       " 'homo_sapiens',\n",
       " 'hucho_hucho',\n",
       " 'ictalurus_punctatus',\n",
       " 'ictidomys_tridecemlineatus',\n",
       " 'jaculus_jaculus',\n",
       " 'junco_hyemalis',\n",
       " 'kryptolebias_marmoratus',\n",
       " 'labrus_bergylta',\n",
       " 'larimichthys_crocea',\n",
       " 'lates_calcarifer',\n",
       " 'laticauda_laticaudata',\n",
       " 'latimeria_chalumnae',\n",
       " 'lepidothrix_coronata',\n",
       " 'lepisosteus_oculatus',\n",
       " 'leptobrachium_leishanense',\n",
       " 'lonchura_striata_domestica',\n",
       " 'loxodonta_africana',\n",
       " 'lynx_canadensis',\n",
       " 'macaca_fascicularis',\n",
       " 'macaca_mulatta',\n",
       " 'macaca_nemestrina',\n",
       " 'malurus_cyaneus_samueli',\n",
       " 'manacus_vitellinus',\n",
       " 'mandrillus_leucophaeus',\n",
       " 'marmota_marmota_marmota',\n",
       " 'mastacembelus_armatus',\n",
       " 'maylandia_zebra',\n",
       " 'meleagris_gallopavo',\n",
       " 'melopsittacus_undulatus',\n",
       " 'meriones_unguiculatus',\n",
       " 'mesocricetus_auratus',\n",
       " 'microcebus_murinus',\n",
       " 'microtus_ochrogaster',\n",
       " 'mola_mola',\n",
       " 'monodelphis_domestica',\n",
       " 'monodon_monoceros',\n",
       " 'monopterus_albus',\n",
       " 'moschus_moschiferus',\n",
       " 'mus_caroli',\n",
       " 'mus_musculus',\n",
       " 'mus_musculus_129s1svimj',\n",
       " 'mus_musculus_aj',\n",
       " 'mus_musculus_akrj',\n",
       " 'mus_musculus_balbcj',\n",
       " 'mus_musculus_c3hhej',\n",
       " 'mus_musculus_c57bl6nj',\n",
       " 'mus_musculus_casteij',\n",
       " 'mus_musculus_cbaj',\n",
       " 'mus_musculus_dba2j',\n",
       " 'mus_musculus_fvbnj',\n",
       " 'mus_musculus_lpj',\n",
       " 'mus_musculus_nodshiltj',\n",
       " 'mus_musculus_nzohlltj',\n",
       " 'mus_musculus_pwkphj',\n",
       " 'mus_musculus_wsbeij',\n",
       " 'mus_pahari',\n",
       " 'mus_spicilegus',\n",
       " 'mus_spretus',\n",
       " 'mustela_putorius_furo',\n",
       " 'myotis_lucifugus',\n",
       " 'myripristis_murdjan',\n",
       " 'naja_naja',\n",
       " 'nannospalax_galili',\n",
       " 'neogobius_melanostomus',\n",
       " 'neolamprologus_brichardi',\n",
       " 'neovison_vison',\n",
       " 'nomascus_leucogenys',\n",
       " 'notamacropus_eugenii',\n",
       " 'notechis_scutatus',\n",
       " 'nothobranchius_furzeri',\n",
       " 'nothoprocta_perdicaria',\n",
       " 'numida_meleagris',\n",
       " 'ochotona_princeps',\n",
       " 'octodon_degus',\n",
       " 'oncorhynchus_kisutch',\n",
       " 'oncorhynchus_mykiss',\n",
       " 'oncorhynchus_tshawytscha',\n",
       " 'oreochromis_aureus',\n",
       " 'oreochromis_niloticus',\n",
       " 'ornithorhynchus_anatinus',\n",
       " 'oryctolagus_cuniculus',\n",
       " 'oryzias_javanicus',\n",
       " 'oryzias_latipes',\n",
       " 'oryzias_latipes_hni',\n",
       " 'oryzias_latipes_hsok',\n",
       " 'oryzias_melastigma',\n",
       " 'oryzias_sinensis',\n",
       " 'otolemur_garnettii',\n",
       " 'otus_sunia',\n",
       " 'ovis_aries',\n",
       " 'ovis_aries_rambouillet',\n",
       " 'pan_paniscus',\n",
       " 'pan_troglodytes',\n",
       " 'panthera_leo',\n",
       " 'panthera_pardus',\n",
       " 'panthera_tigris_altaica',\n",
       " 'papio_anubis',\n",
       " 'parambassis_ranga',\n",
       " 'paramormyrops_kingsleyae',\n",
       " 'parus_major',\n",
       " 'pavo_cristatus',\n",
       " 'pelodiscus_sinensis',\n",
       " 'pelusios_castaneus',\n",
       " 'periophthalmus_magnuspinnatus',\n",
       " 'peromyscus_maniculatus_bairdii',\n",
       " 'petromyzon_marinus',\n",
       " 'phascolarctos_cinereus',\n",
       " 'phasianus_colchicus',\n",
       " 'phocoena_sinus',\n",
       " 'physeter_catodon',\n",
       " 'piliocolobus_tephrosceles',\n",
       " 'podarcis_muralis',\n",
       " 'poecilia_formosa',\n",
       " 'poecilia_latipinna',\n",
       " 'poecilia_mexicana',\n",
       " 'poecilia_reticulata',\n",
       " 'pogona_vitticeps',\n",
       " 'pongo_abelii',\n",
       " 'procavia_capensis',\n",
       " 'prolemur_simus',\n",
       " 'propithecus_coquereli',\n",
       " 'pseudonaja_textilis',\n",
       " 'pteropus_vampyrus',\n",
       " 'pundamilia_nyererei',\n",
       " 'pygocentrus_nattereri',\n",
       " 'rattus_norvegicus',\n",
       " 'rhinolophus_ferrumequinum',\n",
       " 'rhinopithecus_bieti',\n",
       " 'rhinopithecus_roxellana',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'saimiri_boliviensis_boliviensis',\n",
       " 'salarias_fasciatus',\n",
       " 'salmo_salar',\n",
       " 'salmo_trutta',\n",
       " 'salvator_merianae',\n",
       " 'sander_lucioperca',\n",
       " 'sarcophilus_harrisii',\n",
       " 'sciurus_vulgaris',\n",
       " 'scleropages_formosus',\n",
       " 'scophthalmus_maximus',\n",
       " 'serinus_canaria',\n",
       " 'seriola_dumerili',\n",
       " 'seriola_lalandi_dorsalis',\n",
       " 'sinocyclocheilus_anshuiensis',\n",
       " 'sinocyclocheilus_grahami',\n",
       " 'sinocyclocheilus_rhinocerous',\n",
       " 'sorex_araneus',\n",
       " 'sparus_aurata',\n",
       " 'spermophilus_dauricus',\n",
       " 'sphaeramia_orbicularis',\n",
       " 'sphenodon_punctatus',\n",
       " 'stachyris_ruficeps',\n",
       " 'stegastes_partitus',\n",
       " 'strigops_habroptila',\n",
       " 'strix_occidentalis_caurina',\n",
       " 'struthio_camelus_australis',\n",
       " 'suricata_suricatta',\n",
       " 'sus_scrofa',\n",
       " 'sus_scrofa_bamei',\n",
       " 'sus_scrofa_berkshire',\n",
       " 'sus_scrofa_hampshire',\n",
       " 'sus_scrofa_jinhua',\n",
       " 'sus_scrofa_landrace',\n",
       " 'sus_scrofa_largewhite',\n",
       " 'sus_scrofa_meishan',\n",
       " 'sus_scrofa_pietrain',\n",
       " 'sus_scrofa_rongchang',\n",
       " 'sus_scrofa_tibetan',\n",
       " 'sus_scrofa_usmarc',\n",
       " 'sus_scrofa_wuzhishan',\n",
       " 'taeniopygia_guttata',\n",
       " 'takifugu_rubripes',\n",
       " 'terrapene_carolina_triunguis',\n",
       " 'tetraodon_nigroviridis',\n",
       " 'theropithecus_gelada',\n",
       " 'tupaia_belangeri',\n",
       " 'tursiops_truncatus',\n",
       " 'urocitellus_parryii',\n",
       " 'ursus_americanus',\n",
       " 'ursus_maritimus',\n",
       " 'ursus_thibetanus_thibetanus',\n",
       " 'varanus_komodoensis',\n",
       " 'vicugna_pacos',\n",
       " 'vombatus_ursinus',\n",
       " 'vulpes_vulpes',\n",
       " 'xenopus_tropicalis',\n",
       " 'xiphophorus_couchianus',\n",
       " 'xiphophorus_maculatus',\n",
       " 'zalophus_californianus',\n",
       " 'zonotrichia_albicollis',\n",
       " 'zosterops_lateralis_melanops']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if assemblies in orthology table correspond to assemblies in Ensembl db\n",
    "species_in_db_release = gget.ref(species=None, list_species=True, release=109) # list all species in release\n",
    "species_in_db_release\n",
    "#species_names = [x for x in species_names if x in species_in_db_release] # intersect of the two lists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4695f78",
   "metadata": {},
   "source": [
    "```\n",
    "ovis_aries = \"ovis_aries\"\n",
    "ovis_aries_score = 0\n",
    "ovis_aries_rambouillet = \"ovis_aries_rambouillet\"\n",
    "ovis_aries_rambouillet_score = 0\n",
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "ovis_aries = orthology_table_df[\"ovis_aries\"].values.tolist()\n",
    "ovis_aries = [x for x in ovis_aries if str(x) != \"nan\"]\n",
    "for gene in ovis_aries[:100]:\n",
    "\n",
    "    ext = f\"/lookup/id/{gene}\"\n",
    "    response = requests.get(server+ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    response = response.json()[\"species\"]\n",
    "    if response == ovis_aries:\n",
    "        ovis_aries_score += 1\n",
    "    elif response == ovis_aries_rambouillet:\n",
    "        ovis_aries_rambouillet_score += 1\n",
    "    else:\n",
    "        print(response)\n",
    "        \n",
    "print(f\"{ovis_aries}: {ovis_aries_score}\")\n",
    "print(f\"{ovis_aries_rambouillet}: {ovis_aries_rambouillet_score}\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ed182a",
   "metadata": {},
   "source": [
    "```\n",
    "chok1gshd = \"cricetulus_griseus_chok1gshd\"\n",
    "chok1gshd_score = 0\n",
    "crigri = \"cricetulus_griseus_crigri\"\n",
    "crigri_score = 0\n",
    "picr = \"cricetulus_griseus_picr\"\n",
    "picr_score = 0\n",
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "chinese_hamster_table_genes = orthology_table_df[\"Cricetulus griseus\"].values.tolist()\n",
    "chinese_hamster_table_genes = [x for x in chinese_hamster_table_genes if str(x) != \"nan\"]\n",
    "for gene in chinese_hamster_table_genes:\n",
    "\n",
    "    ext = f\"/lookup/id/{gene}\"\n",
    "    response = requests.get(server+ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    response = response.json()[\"species\"]\n",
    "    if response == chok1gshd:\n",
    "        chok1gshd_score += 1\n",
    "    elif response == crigri:\n",
    "        crigri_score += 1\n",
    "    elif reponse == picr:\n",
    "        picr_score += 1\n",
    "    else:\n",
    "        print(reponse)\n",
    "        \n",
    "print(f\"{chok1gshd}: {chok1gshd_score}\")\n",
    "print(f\"{crigri}: {crigri_score}\")\n",
    "print(f\"{picr}: {picr_score}\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "509efca4",
   "metadata": {},
   "source": [
    "```\n",
    "# TODO find naked mole rat assembly\n",
    "female = \"heterocephalus_glaber_female\"\n",
    "female_score = 0\n",
    "male = \"heterocephalus_glaber_male\"\n",
    "male_score = 0\n",
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "\n",
    "naked_mole_rat_table_genes = orthology_table_df[\"Heterocephalus glaber\"].values.tolist()\n",
    "naked_mole_rat_table_genes = [x for x in naked_mole_rat_table_genes if str(x) != \"nan\"]\n",
    "for i, gene in enumerate(naked_mole_rat_table_genes):\n",
    "\n",
    "    ext = f\"/lookup/id/{gene}\"\n",
    "    response = requests.get(server+ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    response = response.json()[\"species\"]\n",
    "    if response == female:\n",
    "        female_score += 1\n",
    "    elif response == male:\n",
    "        male_score += 1\n",
    "    else:\n",
    "        print(reponse)\n",
    "        \n",
    "    if (i % 500) == 0:\n",
    "        print(f\"{female}: {female_score}\")\n",
    "        print(f\"{male}: {male_score}\")\n",
    "        \n",
    "print(f\"{female}: {female_score}\")\n",
    "print(f\"{male}: {male_score}\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "831069f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_genome_annotation_downloaded(organism_name: str):\n",
    "    organism_genome_annotation = downloaded_genomes_dir / (organism_name + \".gtf\")\n",
    "    return organism_genome_annotation.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8bdcf1b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def is_genome_annotation_unziped(organism_name: str):\n",
    "#     organism_genome_annotation = downloaded_genomes_dir / (organism_name + \".gtf\")\n",
    "#     return organism_genome_annotation.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4b26768d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_genome_annotation(organism_name: str):\n",
    "    # get genome annotation gtf\n",
    "    gtf_ftp = gget.ref(organism_name, which=[\"gtf\"], release=ensembl_release)[organism_name][\"annotation_gtf\"][\"ftp\"]\n",
    "    print(gtf_ftp)\n",
    "    \n",
    "    # download gtf genome nanotation\n",
    "    genome_file = downloaded_genomes_dir / (organism_name + \".gtf.gz\")\n",
    "    urllib.request.urlretrieve(gtf_ftp, genome_file)\n",
    "    \n",
    "    # decompress the genome annotation\n",
    "    !gzip -d $genome_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "26e84302",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_annotation(organism_name: str):\n",
    "    if is_genome_annotation_downloaded(organism_name):\n",
    "        print(\"Genome annotation already downloaded and decompressed!\")\n",
    "    else:\n",
    "        print(\"Downloading and then decompressing the genome annotation!\")\n",
    "        download_genome_annotation(organism_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f9822de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "homo_sapiens\n",
      "Genome annotation already downloaded and decompressed!\n",
      "cricetulus_griseus_chok1gshd\n",
      "Genome annotation already downloaded and decompressed!\n",
      "mesocricetus_auratus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "rattus_norvegicus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "cavia_porcellus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "octodon_degus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "heterocephalus_glaber_female\n",
      "Genome annotation already downloaded and decompressed!\n",
      "maylandia_zebra\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pteropus_vampyrus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "monodelphis_domestica\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pelodiscus_sinensis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "anolis_carolinensis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pundamilia_nyererei\n",
      "Genome annotation already downloaded and decompressed!\n",
      "vicugna_pacos\n",
      "Genome annotation already downloaded and decompressed!\n",
      "otolemur_garnettii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "takifugu_rubripes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "neolamprologus_brichardi\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chinchilla_lanigera\n",
      "Genome annotation already downloaded and decompressed!\n",
      "sorex_araneus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ictidomys_tridecemlineatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "geospiza_fortis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "jaculus_jaculus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "myotis_lucifugus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "taeniopygia_guttata\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ficedula_albicollis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chlorocebus_sabaeus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "nomascus_leucogenys\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gasterosteus_aculeatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "petromyzon_marinus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "latimeria_chalumnae\n",
      "Genome annotation already downloaded and decompressed!\n",
      "lepisosteus_oculatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "danio_rerio\n",
      "Genome annotation already downloaded and decompressed!\n",
      "microtus_ochrogaster\n",
      "Genome annotation already downloaded and decompressed!\n",
      "astyanax_mexicanus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gadus_morhua\n",
      "Genome annotation already downloaded and decompressed!\n",
      "xiphophorus_maculatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oryzias_latipes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oreochromis_niloticus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "haplochromis_burtoni\n",
      "Genome annotation already downloaded and decompressed!\n",
      "xenopus_tropicalis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "chrysemys_picta_bellii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gallus_gallus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "meleagris_gallopavo\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ornithorhynchus_anatinus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "sarcophilus_harrisii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "notamacropus_eugenii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "dasypus_novemcinctus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "erinaceus_europaeus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "echinops_telfairi\n",
      "Genome annotation already downloaded and decompressed!\n",
      "callithrix_jacchus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "macaca_fascicularis\n",
      "Genome annotation already downloaded and decompressed!\n",
      "macaca_mulatta\n",
      "Genome annotation already downloaded and decompressed!\n",
      "gorilla_gorilla\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pan_troglodytes\n",
      "Genome annotation already downloaded and decompressed!\n",
      "pongo_abelii\n",
      "Genome annotation already downloaded and decompressed!\n",
      "canis_lupus_familiaris\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ailuropoda_melanoleuca\n",
      "Genome annotation already downloaded and decompressed!\n",
      "mustela_putorius_furo\n",
      "Genome annotation already downloaded and decompressed!\n",
      "felis_catus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "tursiops_truncatus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "loxodonta_africana\n",
      "Genome annotation already downloaded and decompressed!\n",
      "equus_caballus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "bos_taurus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "capra_hircus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "ovis_aries_rambouillet\n",
      "Downloading and then decompressing the genome annotation!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mon Feb 20 20:52:13 2023 INFO Fetching reference information for ovis_aries_rambouillet from Ensembl release: 109.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://ftp.ensembl.org/pub/release-109/gtf/ovis_aries_rambouillet/Ovis_aries_rambouillet.Oar_rambouillet_v1.0.109.gtf.gz\n",
      "ochotona_princeps\n",
      "Genome annotation already downloaded and decompressed!\n",
      "oryctolagus_cuniculus\n",
      "Genome annotation already downloaded and decompressed!\n",
      "tetraodon_nigroviridis\n",
      "Genome annotation already downloaded and decompressed!\n"
     ]
    }
   ],
   "source": [
    "# download assemblies one by one\n",
    "for species in species_names:\n",
    "    print(species)\n",
    "    get_genome_annotation(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c23aad0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gtf_to_bed(organism_name):\n",
    "    # convert gtf to bed format\n",
    "    gtf_file = downloaded_genomes_dir / (organism_name + \".gtf\")\n",
    "    bed_file = downloaded_genomes_dir / (\"sorted_\" + organism_name + \".bed\")\n",
    "    !gtf2bed < $gtf_file > $bed_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "83e394b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "homo_sapiens\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "cricetulus_griseus_chok1gshd\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "mesocricetus_auratus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "rattus_norvegicus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "cavia_porcellus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "octodon_degus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "heterocephalus_glaber_female\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "maylandia_zebra\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pteropus_vampyrus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "monodelphis_domestica\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pelodiscus_sinensis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "anolis_carolinensis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pundamilia_nyererei\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "vicugna_pacos\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "otolemur_garnettii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "takifugu_rubripes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "neolamprologus_brichardi\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "chinchilla_lanigera\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "sorex_araneus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ictidomys_tridecemlineatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "geospiza_fortis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "jaculus_jaculus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "myotis_lucifugus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "taeniopygia_guttata\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ficedula_albicollis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chlorocebus_sabaeus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "nomascus_leucogenys\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gasterosteus_aculeatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "petromyzon_marinus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "latimeria_chalumnae\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "lepisosteus_oculatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "danio_rerio\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "microtus_ochrogaster\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "astyanax_mexicanus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gadus_morhua\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "xiphophorus_maculatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oryzias_latipes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oreochromis_niloticus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "haplochromis_burtoni\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "xenopus_tropicalis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "chrysemys_picta_bellii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gallus_gallus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "meleagris_gallopavo\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ornithorhynchus_anatinus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "sarcophilus_harrisii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "notamacropus_eugenii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "dasypus_novemcinctus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "erinaceus_europaeus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "echinops_telfairi\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "callithrix_jacchus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "macaca_fascicularis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "macaca_mulatta\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "gorilla_gorilla\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pan_troglodytes\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "pongo_abelii\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "canis_lupus_familiaris\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ailuropoda_melanoleuca\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "mustela_putorius_furo\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "felis_catus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "tursiops_truncatus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "loxodonta_africana\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "equus_caballus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "bos_taurus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "capra_hircus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ovis_aries_rambouillet\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "ochotona_princeps\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "oryctolagus_cuniculus\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n",
      "tetraodon_nigroviridis\n",
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n",
      "Warning: Potentially missing gene or transcript ID from GTF attributes (malformed GTF at line [1]?)\n"
     ]
    }
   ],
   "source": [
    "# generate sorted bed files from genome annotations\n",
    "for species in species_names:\n",
    "    print(species)\n",
    "    gtf_to_bed(species)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
