{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e0b6d5b",
   "metadata": {},
   "source": [
    "# Identify hidden genes using the orthology table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac4b68e0",
   "metadata": {},
   "source": [
    "Distinguish hidden gene from a missing gene: a hidden gene is present in at least one organism of a class, but not in all of them. Prioritize (sort) them based on the fullness of their orthology group.\n",
    "\n",
    "**Input:** Orthology table (.xsl, .csv, .tsv), version of Ensembl compara used, taxa of hidden genes, output filename\n",
    "\n",
    "**Output:** A JSON-formatted file with prioritized hidden genes (species with hidden gene, species with ortholog, ortholog ensembl id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a011102",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import sys\n",
    "import json\n",
    "import seaborn as sns\n",
    "import gget\n",
    "from Bio.SeqUtils import GC\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95020f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_api = \"https://rest.ensembl.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1d8e96cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter desired ensembl release\n",
    "ensemble_release = 109"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3791e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter path to orthology table\n",
    "orthology_table_file = pathlib.Path(\"/home/vecerkok/orthology_table/orthology_table-68species_ensembl_ids.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c2356e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter taxa of hidden genes\n",
    "taxon = \"Aves\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "014dee6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter output filename\n",
    "hidden_genes_file = pathlib.Path(\"100_avian_genes_to_search\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7f2c58c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enter the fraction of genes to be output\n",
    "fraction_output_genes = 0.00185"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b37c6595",
   "metadata": {},
   "source": [
    "## Orthology table preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e4316f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthology_table(table_file: str):\n",
    "    file_type = table_file.suffix\n",
    "    \n",
    "    if file_type == \".csv\":\n",
    "        return pd.read_csv(table_file, sep=\",\", header=0, index_col=0, low_memory=False)\n",
    "    elif file_type == \".tsv\":\n",
    "        return pd.read_csv(table_file, sep=\"\\t\", header=0, index_col=0, low_memory=False)\n",
    "    elif (file_type == \".xsls\") or (file_type == \".xsl\"):\n",
    "        return pd.read_excel(table_file, header=0, index_col=0, low_memory=False)\n",
    "    else:\n",
    "        print(\"Accepted file formats: .csv, .tsv, .xsls, .xsl!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "71a78b87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homo sapiens</th>\n",
       "      <th>Cricetulus griseus</th>\n",
       "      <th>Mesocricetus auratus</th>\n",
       "      <th>Rattus norvegicus</th>\n",
       "      <th>Cavia porcellus</th>\n",
       "      <th>Octodon degus</th>\n",
       "      <th>Heterocephalus glaber</th>\n",
       "      <th>Maylandia zebra</th>\n",
       "      <th>Pteropus vampyrus</th>\n",
       "      <th>Monodelphis domestica</th>\n",
       "      <th>...</th>\n",
       "      <th>Felis catus</th>\n",
       "      <th>Tursiops truncatus</th>\n",
       "      <th>Loxodonta africana</th>\n",
       "      <th>Equus caballus</th>\n",
       "      <th>Bos taurus</th>\n",
       "      <th>Capra hircus</th>\n",
       "      <th>Ovis aries</th>\n",
       "      <th>Ochotona princeps</th>\n",
       "      <th>Oryctolagus cuniculus</th>\n",
       "      <th>Tetraodon nigroviridis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000167103</td>\n",
       "      <td>ENSCGRG00001009175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSRNOG00000048676</td>\n",
       "      <td>ENSCPOG00000005283</td>\n",
       "      <td>ENSODEG00000003703</td>\n",
       "      <td>ENSHGLG00000018036</td>\n",
       "      <td>ENSMZEG00005000701</td>\n",
       "      <td>ENSPVAG00000000967</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSFCAG00000001492</td>\n",
       "      <td>ENSTTRG00000016470</td>\n",
       "      <td>ENSLAFG00000000585</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSBTAG00000003387</td>\n",
       "      <td>ENSCHIG00000020199</td>\n",
       "      <td>ENSOARG00020015289</td>\n",
       "      <td>ENSOPRG00000016354</td>\n",
       "      <td>ENSOCUG00000026071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000258873</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPVAG00000003020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTTRG00000014200</td>\n",
       "      <td>ENSLAFG00000030167</td>\n",
       "      <td>ENSECAG00000043636</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOPRG00000000440</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>ENSCGRG00001021594</td>\n",
       "      <td>ENSMAUG00000015131</td>\n",
       "      <td>ENSRNOG00000069330</td>\n",
       "      <td>ENSCPOG00000024213</td>\n",
       "      <td>ENSODEG00000009370</td>\n",
       "      <td>ENSHGLG00000006136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMODG00000042514</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSFCAG00000018619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSLAFG00000031828</td>\n",
       "      <td>ENSECAG00000000874</td>\n",
       "      <td>ENSBTAG00000003220</td>\n",
       "      <td>ENSCHIG00000003089</td>\n",
       "      <td>ENSOARG00020009415</td>\n",
       "      <td>ENSOPRG00000018458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000185245</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSRNOG00000025959</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPVAG00000003789</td>\n",
       "      <td>ENSMODG00000005365</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSFCAG00000008042</td>\n",
       "      <td>ENSTTRG00000001296</td>\n",
       "      <td>ENSLAFG00000032715</td>\n",
       "      <td>ENSECAG00000040351</td>\n",
       "      <td>ENSBTAG00000004909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOARG00020025177</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOCUG00000001099</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198870</td>\n",
       "      <td>ENSCGRG00001004554</td>\n",
       "      <td>ENSMAUG00000013976</td>\n",
       "      <td>ENSRNOG00000027911</td>\n",
       "      <td>ENSCPOG00000021170</td>\n",
       "      <td>ENSODEG00000015481</td>\n",
       "      <td>ENSHGLG00000016873</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPVAG00000001428</td>\n",
       "      <td>ENSMODG00000012800</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSFCAG00000003766</td>\n",
       "      <td>ENSTTRG00000007671</td>\n",
       "      <td>ENSLAFG00000008238</td>\n",
       "      <td>ENSECAG00000022240</td>\n",
       "      <td>ENSBTAG00000004140</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOARG00020002366</td>\n",
       "      <td>ENSOPRG00000001115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54115</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54116</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTNIG00000020644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54117</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54118</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTNIG00000020257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54119</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMZEG00005027689</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54120 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Homo sapiens  Cricetulus griseus Mesocricetus auratus  \\\n",
       "0      ENSG00000167103  ENSCGRG00001009175                  NaN   \n",
       "1      ENSG00000258873                 NaN                  NaN   \n",
       "2      ENSG00000213088  ENSCGRG00001021594   ENSMAUG00000015131   \n",
       "3      ENSG00000185245                 NaN                  NaN   \n",
       "4      ENSG00000198870  ENSCGRG00001004554   ENSMAUG00000013976   \n",
       "...                ...                 ...                  ...   \n",
       "54115              NaN                 NaN                  NaN   \n",
       "54116              NaN                 NaN                  NaN   \n",
       "54117              NaN                 NaN                  NaN   \n",
       "54118              NaN                 NaN                  NaN   \n",
       "54119              NaN                 NaN                  NaN   \n",
       "\n",
       "        Rattus norvegicus     Cavia porcellus       Octodon degus  \\\n",
       "0      ENSRNOG00000048676  ENSCPOG00000005283  ENSODEG00000003703   \n",
       "1                     NaN                 NaN                 NaN   \n",
       "2      ENSRNOG00000069330  ENSCPOG00000024213  ENSODEG00000009370   \n",
       "3      ENSRNOG00000025959                 NaN                 NaN   \n",
       "4      ENSRNOG00000027911  ENSCPOG00000021170  ENSODEG00000015481   \n",
       "...                   ...                 ...                 ...   \n",
       "54115                 NaN                 NaN                 NaN   \n",
       "54116                 NaN                 NaN                 NaN   \n",
       "54117                 NaN                 NaN                 NaN   \n",
       "54118                 NaN                 NaN                 NaN   \n",
       "54119                 NaN                 NaN                 NaN   \n",
       "\n",
       "      Heterocephalus glaber     Maylandia zebra   Pteropus vampyrus  \\\n",
       "0        ENSHGLG00000018036  ENSMZEG00005000701  ENSPVAG00000000967   \n",
       "1                       NaN                 NaN  ENSPVAG00000003020   \n",
       "2        ENSHGLG00000006136                 NaN                 NaN   \n",
       "3                       NaN                 NaN  ENSPVAG00000003789   \n",
       "4        ENSHGLG00000016873                 NaN  ENSPVAG00000001428   \n",
       "...                     ...                 ...                 ...   \n",
       "54115                   NaN                 NaN                 NaN   \n",
       "54116                   NaN                 NaN                 NaN   \n",
       "54117                   NaN                 NaN                 NaN   \n",
       "54118                   NaN                 NaN                 NaN   \n",
       "54119                   NaN  ENSMZEG00005027689                 NaN   \n",
       "\n",
       "      Monodelphis domestica  ...         Felis catus  Tursiops truncatus  \\\n",
       "0                       NaN  ...  ENSFCAG00000001492  ENSTTRG00000016470   \n",
       "1                       NaN  ...                 NaN  ENSTTRG00000014200   \n",
       "2        ENSMODG00000042514  ...  ENSFCAG00000018619                 NaN   \n",
       "3        ENSMODG00000005365  ...  ENSFCAG00000008042  ENSTTRG00000001296   \n",
       "4        ENSMODG00000012800  ...  ENSFCAG00000003766  ENSTTRG00000007671   \n",
       "...                     ...  ...                 ...                 ...   \n",
       "54115                   NaN  ...                 NaN                 NaN   \n",
       "54116                   NaN  ...                 NaN                 NaN   \n",
       "54117                   NaN  ...                 NaN                 NaN   \n",
       "54118                   NaN  ...                 NaN                 NaN   \n",
       "54119                   NaN  ...                 NaN                 NaN   \n",
       "\n",
       "       Loxodonta africana      Equus caballus          Bos taurus  \\\n",
       "0      ENSLAFG00000000585                 NaN  ENSBTAG00000003387   \n",
       "1      ENSLAFG00000030167  ENSECAG00000043636                 NaN   \n",
       "2      ENSLAFG00000031828  ENSECAG00000000874  ENSBTAG00000003220   \n",
       "3      ENSLAFG00000032715  ENSECAG00000040351  ENSBTAG00000004909   \n",
       "4      ENSLAFG00000008238  ENSECAG00000022240  ENSBTAG00000004140   \n",
       "...                   ...                 ...                 ...   \n",
       "54115                 NaN                 NaN                 NaN   \n",
       "54116                 NaN                 NaN                 NaN   \n",
       "54117                 NaN                 NaN                 NaN   \n",
       "54118                 NaN                 NaN                 NaN   \n",
       "54119                 NaN                 NaN                 NaN   \n",
       "\n",
       "             Capra hircus          Ovis aries   Ochotona princeps  \\\n",
       "0      ENSCHIG00000020199  ENSOARG00020015289  ENSOPRG00000016354   \n",
       "1                     NaN                 NaN  ENSOPRG00000000440   \n",
       "2      ENSCHIG00000003089  ENSOARG00020009415  ENSOPRG00000018458   \n",
       "3                     NaN  ENSOARG00020025177                 NaN   \n",
       "4                     NaN  ENSOARG00020002366  ENSOPRG00000001115   \n",
       "...                   ...                 ...                 ...   \n",
       "54115                 NaN                 NaN                 NaN   \n",
       "54116                 NaN                 NaN                 NaN   \n",
       "54117                 NaN                 NaN                 NaN   \n",
       "54118                 NaN                 NaN                 NaN   \n",
       "54119                 NaN                 NaN                 NaN   \n",
       "\n",
       "      Oryctolagus cuniculus Tetraodon nigroviridis  \n",
       "0        ENSOCUG00000026071                    NaN  \n",
       "1                       NaN                    NaN  \n",
       "2                       NaN                    NaN  \n",
       "3        ENSOCUG00000001099                    NaN  \n",
       "4                       NaN                    NaN  \n",
       "...                     ...                    ...  \n",
       "54115                   NaN                    NaN  \n",
       "54116                   NaN     ENSTNIG00000020644  \n",
       "54117                   NaN                    NaN  \n",
       "54118                   NaN     ENSTNIG00000020257  \n",
       "54119                   NaN                    NaN  \n",
       "\n",
       "[54120 rows x 68 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# upload the orthology table\n",
    "orthology_df = get_orthology_table(orthology_table_file)\n",
    "orthology_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a864aa91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_list_organisms_orthology_table(orth_df: pd.DataFrame):\n",
    "    # get a list column labels\n",
    "    column_names = orth_df.columns.tolist()\n",
    "    # convert species name to lowercase (useful for ensemble rest api)\n",
    "    organism_names = [x.lower().replace(\" \", \"_\") for x in column_names]\n",
    "    organism_names = list(map(lambda x: x.replace(\"heterocephalus_glaber\", \"heterocephalus_glaber_female\"), organism_names))\n",
    "    organism_names = list(map(lambda x: x.replace(\"gorilla_gorilla_gorilla\", \"gorilla_gorilla\"), organism_names))\n",
    "    organism_names = list(map(lambda x: x.replace(\"cricetulus_griseus\", \"cricetulus_griseus_chok1gshd\"), organism_names))\n",
    "    organism_names = list(map(lambda x: x.replace(\"ovis_aries\", \"ovis_aries_rambouillet\"), organism_names))\n",
    "    orth_df.columns = organism_names\n",
    "    return orth_df, organism_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7d08c4c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transpose_orthology_table(orth_df: pd.DataFrame):\n",
    "    return orth_df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "496a1d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_organism_class(organism: str):\n",
    "    # get taxonomy classification information for organism from Ensembl rest api\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = f\"/taxonomy/classification/{organism}?\"\n",
    "\n",
    "    response = requests.get(server+ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    response = json.dumps(response.json())\n",
    "    \n",
    "    # get class from taxonomy info\n",
    "    classes = [\"Mammalia\", \"Aves\", \"Reptilia\", \"Actinopteri\", \"Amphibia\"]\n",
    "    for cl in classes:\n",
    "        if (cl in response):\n",
    "            return cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7cb99a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add class info to the orthology table\n",
    "def add_classes_to_orthology_database(orth_df: pd.DataFrame):\n",
    "    # get a list of species\n",
    "    orth_df, species = get_list_organisms_orthology_table(orth_df)\n",
    "    # transpose the dataframe\n",
    "    orth_df = transpose_orthology_table(orth_df)\n",
    "    \n",
    "    # assign species classes\n",
    "    for specie in species:\n",
    "        orth_df.loc[specie, \"class\"] = get_organism_class(specie)\n",
    "    return orth_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cdcb4803",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>54111</th>\n",
       "      <th>54112</th>\n",
       "      <th>54113</th>\n",
       "      <th>54114</th>\n",
       "      <th>54115</th>\n",
       "      <th>54116</th>\n",
       "      <th>54117</th>\n",
       "      <th>54118</th>\n",
       "      <th>54119</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>homo_sapiens</th>\n",
       "      <td>ENSG00000167103</td>\n",
       "      <td>ENSG00000258873</td>\n",
       "      <td>ENSG00000213088</td>\n",
       "      <td>ENSG00000185245</td>\n",
       "      <td>ENSG00000198870</td>\n",
       "      <td>ENSG00000256553</td>\n",
       "      <td>ENSG00000255569</td>\n",
       "      <td>ENSG00000120594</td>\n",
       "      <td>ENSG00000161381</td>\n",
       "      <td>ENSG00000100557</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cricetulus_griseus_chok1gshd</th>\n",
       "      <td>ENSCGRG00001009175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCGRG00001021594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCGRG00001004554</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCGRG00001024180</td>\n",
       "      <td>ENSCGRG00001018272</td>\n",
       "      <td>ENSCGRG00001010407</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mesocricetus_auratus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMAUG00000015131</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMAUG00000013976</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMAUG00000015907</td>\n",
       "      <td>ENSMAUG00000013689</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rattus_norvegicus</th>\n",
       "      <td>ENSRNOG00000048676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSRNOG00000069330</td>\n",
       "      <td>ENSRNOG00000025959</td>\n",
       "      <td>ENSRNOG00000027911</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSRNOG00000000142</td>\n",
       "      <td>ENSRNOG00000021536</td>\n",
       "      <td>ENSRNOG00000062766</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cavia_porcellus</th>\n",
       "      <td>ENSCPOG00000005283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCPOG00000024213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCPOG00000021170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCPOG00000007300</td>\n",
       "      <td>ENSCPOG00000020274</td>\n",
       "      <td>ENSCPOG00000003345</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>capra_hircus</th>\n",
       "      <td>ENSCHIG00000020199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCHIG00000003089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSCHIG00000023670</td>\n",
       "      <td>ENSCHIG00000024239</td>\n",
       "      <td>ENSCHIG00000027108</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ovis_aries_rambouillet</th>\n",
       "      <td>ENSOARG00020015289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOARG00020009415</td>\n",
       "      <td>ENSOARG00020025177</td>\n",
       "      <td>ENSOARG00020002366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOARG00020023004</td>\n",
       "      <td>ENSOARG00020016631</td>\n",
       "      <td>ENSOARG00020020200</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ochotona_princeps</th>\n",
       "      <td>ENSOPRG00000016354</td>\n",
       "      <td>ENSOPRG00000000440</td>\n",
       "      <td>ENSOPRG00000018458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOPRG00000001115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOPRG00000009876</td>\n",
       "      <td>ENSOPRG00000007050</td>\n",
       "      <td>ENSOPRG00000016009</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oryctolagus_cuniculus</th>\n",
       "      <td>ENSOCUG00000026071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOCUG00000001099</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSOCUG00000014482</td>\n",
       "      <td>ENSOCUG00000011309</td>\n",
       "      <td>ENSOCUG00000015402</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tetraodon_nigroviridis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTNIG00000005847</td>\n",
       "      <td>ENSTNIG00000000704</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTNIG00000020644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTNIG00000020257</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Actinopteri</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 54121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               0                   1  \\\n",
       "homo_sapiens                     ENSG00000167103     ENSG00000258873   \n",
       "cricetulus_griseus_chok1gshd  ENSCGRG00001009175                 NaN   \n",
       "mesocricetus_auratus                         NaN                 NaN   \n",
       "rattus_norvegicus             ENSRNOG00000048676                 NaN   \n",
       "cavia_porcellus               ENSCPOG00000005283                 NaN   \n",
       "...                                          ...                 ...   \n",
       "capra_hircus                  ENSCHIG00000020199                 NaN   \n",
       "ovis_aries_rambouillet        ENSOARG00020015289                 NaN   \n",
       "ochotona_princeps             ENSOPRG00000016354  ENSOPRG00000000440   \n",
       "oryctolagus_cuniculus         ENSOCUG00000026071                 NaN   \n",
       "tetraodon_nigroviridis                       NaN                 NaN   \n",
       "\n",
       "                                               2                   3  \\\n",
       "homo_sapiens                     ENSG00000213088     ENSG00000185245   \n",
       "cricetulus_griseus_chok1gshd  ENSCGRG00001021594                 NaN   \n",
       "mesocricetus_auratus          ENSMAUG00000015131                 NaN   \n",
       "rattus_norvegicus             ENSRNOG00000069330  ENSRNOG00000025959   \n",
       "cavia_porcellus               ENSCPOG00000024213                 NaN   \n",
       "...                                          ...                 ...   \n",
       "capra_hircus                  ENSCHIG00000003089                 NaN   \n",
       "ovis_aries_rambouillet        ENSOARG00020009415  ENSOARG00020025177   \n",
       "ochotona_princeps             ENSOPRG00000018458                 NaN   \n",
       "oryctolagus_cuniculus                        NaN  ENSOCUG00000001099   \n",
       "tetraodon_nigroviridis                       NaN                 NaN   \n",
       "\n",
       "                                               4                5  \\\n",
       "homo_sapiens                     ENSG00000198870  ENSG00000256553   \n",
       "cricetulus_griseus_chok1gshd  ENSCGRG00001004554              NaN   \n",
       "mesocricetus_auratus          ENSMAUG00000013976              NaN   \n",
       "rattus_norvegicus             ENSRNOG00000027911              NaN   \n",
       "cavia_porcellus               ENSCPOG00000021170              NaN   \n",
       "...                                          ...              ...   \n",
       "capra_hircus                                 NaN              NaN   \n",
       "ovis_aries_rambouillet        ENSOARG00020002366              NaN   \n",
       "ochotona_princeps             ENSOPRG00000001115              NaN   \n",
       "oryctolagus_cuniculus                        NaN              NaN   \n",
       "tetraodon_nigroviridis                       NaN              NaN   \n",
       "\n",
       "                                            6                   7  \\\n",
       "homo_sapiens                  ENSG00000255569     ENSG00000120594   \n",
       "cricetulus_griseus_chok1gshd              NaN  ENSCGRG00001024180   \n",
       "mesocricetus_auratus                      NaN                 NaN   \n",
       "rattus_norvegicus                         NaN  ENSRNOG00000000142   \n",
       "cavia_porcellus                           NaN  ENSCPOG00000007300   \n",
       "...                                       ...                 ...   \n",
       "capra_hircus                              NaN  ENSCHIG00000023670   \n",
       "ovis_aries_rambouillet                    NaN  ENSOARG00020023004   \n",
       "ochotona_princeps                         NaN  ENSOPRG00000009876   \n",
       "oryctolagus_cuniculus                     NaN  ENSOCUG00000014482   \n",
       "tetraodon_nigroviridis                    NaN  ENSTNIG00000005847   \n",
       "\n",
       "                                               8                   9  ...  \\\n",
       "homo_sapiens                     ENSG00000161381     ENSG00000100557  ...   \n",
       "cricetulus_griseus_chok1gshd  ENSCGRG00001018272  ENSCGRG00001010407  ...   \n",
       "mesocricetus_auratus          ENSMAUG00000015907  ENSMAUG00000013689  ...   \n",
       "rattus_norvegicus             ENSRNOG00000021536  ENSRNOG00000062766  ...   \n",
       "cavia_porcellus               ENSCPOG00000020274  ENSCPOG00000003345  ...   \n",
       "...                                          ...                 ...  ...   \n",
       "capra_hircus                  ENSCHIG00000024239  ENSCHIG00000027108  ...   \n",
       "ovis_aries_rambouillet        ENSOARG00020016631  ENSOARG00020020200  ...   \n",
       "ochotona_princeps             ENSOPRG00000007050  ENSOPRG00000016009  ...   \n",
       "oryctolagus_cuniculus         ENSOCUG00000011309  ENSOCUG00000015402  ...   \n",
       "tetraodon_nigroviridis        ENSTNIG00000000704                 NaN  ...   \n",
       "\n",
       "                             54111 54112 54113 54114 54115  \\\n",
       "homo_sapiens                   NaN   NaN   NaN   NaN   NaN   \n",
       "cricetulus_griseus_chok1gshd   NaN   NaN   NaN   NaN   NaN   \n",
       "mesocricetus_auratus           NaN   NaN   NaN   NaN   NaN   \n",
       "rattus_norvegicus              NaN   NaN   NaN   NaN   NaN   \n",
       "cavia_porcellus                NaN   NaN   NaN   NaN   NaN   \n",
       "...                            ...   ...   ...   ...   ...   \n",
       "capra_hircus                   NaN   NaN   NaN   NaN   NaN   \n",
       "ovis_aries_rambouillet         NaN   NaN   NaN   NaN   NaN   \n",
       "ochotona_princeps              NaN   NaN   NaN   NaN   NaN   \n",
       "oryctolagus_cuniculus          NaN   NaN   NaN   NaN   NaN   \n",
       "tetraodon_nigroviridis         NaN   NaN   NaN   NaN   NaN   \n",
       "\n",
       "                                           54116 54117               54118  \\\n",
       "homo_sapiens                                 NaN   NaN                 NaN   \n",
       "cricetulus_griseus_chok1gshd                 NaN   NaN                 NaN   \n",
       "mesocricetus_auratus                         NaN   NaN                 NaN   \n",
       "rattus_norvegicus                            NaN   NaN                 NaN   \n",
       "cavia_porcellus                              NaN   NaN                 NaN   \n",
       "...                                          ...   ...                 ...   \n",
       "capra_hircus                                 NaN   NaN                 NaN   \n",
       "ovis_aries_rambouillet                       NaN   NaN                 NaN   \n",
       "ochotona_princeps                            NaN   NaN                 NaN   \n",
       "oryctolagus_cuniculus                        NaN   NaN                 NaN   \n",
       "tetraodon_nigroviridis        ENSTNIG00000020644   NaN  ENSTNIG00000020257   \n",
       "\n",
       "                             54119        class  \n",
       "homo_sapiens                   NaN     Mammalia  \n",
       "cricetulus_griseus_chok1gshd   NaN         None  \n",
       "mesocricetus_auratus           NaN     Mammalia  \n",
       "rattus_norvegicus              NaN     Mammalia  \n",
       "cavia_porcellus                NaN     Mammalia  \n",
       "...                            ...          ...  \n",
       "capra_hircus                   NaN     Mammalia  \n",
       "ovis_aries_rambouillet         NaN         None  \n",
       "ochotona_princeps              NaN     Mammalia  \n",
       "oryctolagus_cuniculus          NaN     Mammalia  \n",
       "tetraodon_nigroviridis         NaN  Actinopteri  \n",
       "\n",
       "[68 rows x 54121 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orthology_df = add_classes_to_orthology_database(orthology_df)\n",
    "orthology_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b0fc180",
   "metadata": {},
   "source": [
    "## Identifying hidden genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c4da5e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_occurence(orth_df):\n",
    "    # find percentage of occurence per gene\n",
    "    genes_occurence = pd.DataFrame(orth_df.notna().sum() / len(orth_df), columns=[\"percentage\"])\n",
    "    return genes_occurence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9fbcef23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scoring function: the fuller the orthology group, the higher the gene score (the more likely it is hidden rather than missing)\n",
    "def inititate_score_table(orth_df: pd.DataFrame):\n",
    "    genes = orth_df.drop(columns=[\"class\"]).columns.values.tolist()\n",
    "    starting_scores = np.zeros(len(genes))\n",
    "    score_table = pd.DataFrame({\"scores\": starting_scores}, index=genes)\n",
    "    return score_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7d6336b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add taxa and non taxa occurence percentage to the score sum\n",
    "# argument factor can be used to increase the significance of gene occurance in taxa\n",
    "def add_occurence_score(scores_df: pd.DataFrame, occur_df: pd.DataFrame, factor: float = 1.):\n",
    "    scores_df[\"scores\"] = scores_df[\"scores\"] + occur_df[\"percentage\"]\n",
    "    return scores_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6593c386",
   "metadata": {},
   "outputs": [],
   "source": [
    "# prioritize gene to search for based on score\n",
    "def get_genes_scores(ortho_df: pd.DataFrame):\n",
    "    scores_df = inititate_score_table(ortho_df)\n",
    "    \n",
    "    # get gene occurence percentage in taxon\n",
    "    taxon_df = ortho_df.loc[(ortho_df[\"class\"] == taxon)].drop(columns=\"class\")\n",
    "                            \n",
    "    # get gene occurence outside of taxon\n",
    "    taxon_excluded_df = ortho_df.loc[(ortho_df[\"class\"] != taxon)].drop(columns=\"class\")\n",
    "    \n",
    "    gene_taxon_occurence = get_gene_occurence(taxon_df)\n",
    "    gene_non_taxon_occurence = get_gene_occurence(taxon_excluded_df)\n",
    "    \n",
    "    scores_df = add_occurence_score(scores_df, gene_taxon_occurence)\n",
    "    scores_df = add_occurence_score(scores_df, gene_non_taxon_occurence)\n",
    "\n",
    "    return scores_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "43a309c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.841270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.253968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.987302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.228571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.234921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54115</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54116</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54117</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54118</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54119</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54120 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         scores\n",
       "0      0.841270\n",
       "1      0.253968\n",
       "2      0.987302\n",
       "3      1.228571\n",
       "4      1.234921\n",
       "...         ...\n",
       "54115  0.031746\n",
       "54116  0.031746\n",
       "54117  0.031746\n",
       "54118  0.031746\n",
       "54119  0.031746\n",
       "\n",
       "[54120 rows x 1 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_df = get_genes_scores(orthology_df)\n",
    "score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ac5776dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.841270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.253968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.987302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.228571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.234921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54115</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54116</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54117</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54118</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54119</th>\n",
       "      <td>0.031746</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53952 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         scores\n",
       "0      0.841270\n",
       "1      0.253968\n",
       "2      0.987302\n",
       "3      1.228571\n",
       "4      1.234921\n",
       "...         ...\n",
       "54115  0.031746\n",
       "54116  0.031746\n",
       "54117  0.031746\n",
       "54118  0.031746\n",
       "54119  0.031746\n",
       "\n",
       "[53952 rows x 1 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_genes_with_all_orthologs_present(gene_scores: pd.DataFrame, ortho_df: pd.DataFrame):\n",
    "    is_perfect_score_all_present = False\n",
    "    max_score = gene_scores.values.max()\n",
    "    index_highest_scoring_gene = gene_scores.loc[gene_scores[\"scores\"] == max_score].index[0]\n",
    "    if orthology_df[index_highest_scoring_gene].isna().sum() == 0:\n",
    "        is_perfect_score_all_present = True\n",
    "    # drop all rows with perfect score\n",
    "    gene_scores = gene_scores[gene_scores[\"scores\"] < max_score].copy(deep=True)\n",
    "    return gene_scores\n",
    "\n",
    "filtered_score_df = filter_genes_with_all_orthologs_present(score_df, orthology_df)\n",
    "filtered_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "66cfc910",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6758</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10545</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2150</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6981</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6983</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3438</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8839</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3435</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9054</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9051</th>\n",
       "      <td>1.984127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         scores\n",
       "6758   1.984127\n",
       "10545  1.984127\n",
       "2150   1.984127\n",
       "6981   1.984127\n",
       "6983   1.984127\n",
       "...         ...\n",
       "3438   1.984127\n",
       "8839   1.984127\n",
       "3435   1.984127\n",
       "9054   1.984127\n",
       "9051   1.984127\n",
       "\n",
       "[100 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# return a list of prioritized genes\n",
    "# percentage: 0.1 -> returns 10% best scoring genes\n",
    "# gene index id corresponds\n",
    "def prirotize_gene_search(gene_scores: pd.DataFrame, percentage: float):\n",
    "    max_score = gene_scores.values.max()\n",
    "    gene_scores.sort_values(\"scores\", ascending=False, inplace=True)\n",
    "    number_of_genes = int(score_df.shape[0] * percentage)\n",
    "    \n",
    "    return gene_scores.iloc[:number_of_genes, :]\n",
    "\n",
    "prioritized_score_df = prirotize_gene_search(filtered_score_df, fraction_output_genes)\n",
    "prioritized_score_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "77415f0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('vicugna_pacos', 'homo_sapiens', 'ENSG00000132768'),\n",
       " ('sorex_araneus', 'homo_sapiens', 'ENSG00000104213'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000163710'),\n",
       " ('ochotona_princeps', 'homo_sapiens', 'ENSG00000104447'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000103091'),\n",
       " ('astyanax_mexicanus', 'homo_sapiens', 'ENSG00000139324'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000181722'),\n",
       " ('jaculus_jaculus', 'homo_sapiens', 'ENSG00000125863'),\n",
       " ('dasypus_novemcinctus', 'homo_sapiens', 'ENSG00000170615'),\n",
       " ('tetraodon_nigroviridis', 'homo_sapiens', 'ENSG00000157350'),\n",
       " ('vicugna_pacos', 'homo_sapiens', 'ENSG00000198400'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000122299'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164532'),\n",
       " ('heterocephalus_glaber_female', 'homo_sapiens', 'ENSG00000144747'),\n",
       " ('astyanax_mexicanus', 'homo_sapiens', 'ENSG00000136936'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000142784'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000145087'),\n",
       " ('anolis_carolinensis', 'homo_sapiens', 'ENSG00000069943'),\n",
       " ('notamacropus_eugenii', 'homo_sapiens', 'ENSG00000183826'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000141441'),\n",
       " ('anolis_carolinensis', 'homo_sapiens', 'ENSG00000149090'),\n",
       " ('gasterosteus_aculeatus', 'homo_sapiens', 'ENSG00000113492'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000165606'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164176'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000097096'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000057663'),\n",
       " ('heterocephalus_glaber_female', 'homo_sapiens', 'ENSG00000132823'),\n",
       " ('otolemur_garnettii', 'homo_sapiens', 'ENSG00000115827'),\n",
       " ('otolemur_garnettii', 'homo_sapiens', 'ENSG00000088035'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164933'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000117528'),\n",
       " ('heterocephalus_glaber_female', 'homo_sapiens', 'ENSG00000151806'),\n",
       " ('takifugu_rubripes', 'homo_sapiens', 'ENSG00000100697'),\n",
       " ('maylandia_zebra', 'homo_sapiens', 'ENSG00000136052'),\n",
       " ('anolis_carolinensis', 'homo_sapiens', 'ENSG00000169604'),\n",
       " ('myotis_lucifugus', 'homo_sapiens', 'ENSG00000186575'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164649'),\n",
       " ('otolemur_garnettii', 'homo_sapiens', 'ENSG00000119760'),\n",
       " ('pelodiscus_sinensis', 'homo_sapiens', 'ENSG00000155229'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000090530'),\n",
       " ('pundamilia_nyererei', 'homo_sapiens', 'ENSG00000146918'),\n",
       " ('takifugu_rubripes', 'homo_sapiens', 'ENSG00000100225'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000070718'),\n",
       " ('sorex_araneus', 'homo_sapiens', 'ENSG00000162736'),\n",
       " ('myotis_lucifugus', 'homo_sapiens', 'ENSG00000170775'),\n",
       " ('pelodiscus_sinensis', 'homo_sapiens', 'ENSG00000072952'),\n",
       " ('bos_taurus', 'homo_sapiens', 'ENSG00000129675'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000144320'),\n",
       " ('dasypus_novemcinctus', 'homo_sapiens', 'ENSG00000118873'),\n",
       " ('vicugna_pacos', 'homo_sapiens', 'ENSG00000118965'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000158296'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000159461'),\n",
       " ('otolemur_garnettii', 'homo_sapiens', 'ENSG00000172159'),\n",
       " ('latimeria_chalumnae', 'homo_sapiens', 'ENSG00000115902'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000138303'),\n",
       " ('chlorocebus_sabaeus', 'homo_sapiens', 'ENSG00000092978'),\n",
       " ('gadus_morhua', 'homo_sapiens', 'ENSG00000085276'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164161'),\n",
       " ('vicugna_pacos', 'homo_sapiens', 'ENSG00000176444'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000080815'),\n",
       " ('octodon_degus', 'homo_sapiens', 'ENSG00000196305'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000150630'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000065883'),\n",
       " ('astyanax_mexicanus', 'homo_sapiens', 'ENSG00000131373'),\n",
       " ('macaca_fascicularis', 'homo_sapiens', 'ENSG00000164440'),\n",
       " ('astyanax_mexicanus', 'homo_sapiens', 'ENSG00000115295'),\n",
       " ('otolemur_garnettii', 'homo_sapiens', 'ENSG00000111596'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000172164'),\n",
       " ('notamacropus_eugenii', 'homo_sapiens', 'ENSG00000163558'),\n",
       " ('sarcophilus_harrisii', 'homo_sapiens', 'ENSG00000010818'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000116885'),\n",
       " ('takifugu_rubripes', 'homo_sapiens', 'ENSG00000108349'),\n",
       " ('mustela_putorius_furo', 'homo_sapiens', 'ENSG00000124789'),\n",
       " ('octodon_degus', 'homo_sapiens', 'ENSG00000108423'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000137460'),\n",
       " ('myotis_lucifugus', 'homo_sapiens', 'ENSG00000005175'),\n",
       " ('ochotona_princeps', 'homo_sapiens', 'ENSG00000169946'),\n",
       " ('sorex_araneus', 'homo_sapiens', 'ENSG00000135365'),\n",
       " ('microtus_ochrogaster', 'homo_sapiens', 'ENSG00000155792'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000138614'),\n",
       " ('capra_hircus', 'homo_sapiens', 'ENSG00000113456'),\n",
       " ('sorex_araneus', 'homo_sapiens', 'ENSG00000145882'),\n",
       " ('octodon_degus', 'homo_sapiens', 'ENSG00000108176'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000168216'),\n",
       " ('jaculus_jaculus', 'homo_sapiens', 'ENSG00000186908'),\n",
       " ('maylandia_zebra', 'homo_sapiens', 'ENSG00000138190'),\n",
       " ('mesocricetus_auratus', 'homo_sapiens', 'ENSG00000125124'),\n",
       " ('ornithorhynchus_anatinus', 'homo_sapiens', 'ENSG00000143207'),\n",
       " ('xenopus_tropicalis', 'homo_sapiens', 'ENSG00000155189'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000110427'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000134852'),\n",
       " ('notamacropus_eugenii', 'homo_sapiens', 'ENSG00000061987'),\n",
       " ('latimeria_chalumnae', 'homo_sapiens', 'ENSG00000091844'),\n",
       " ('tetraodon_nigroviridis', 'homo_sapiens', 'ENSG00000126016'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000164144'),\n",
       " ('maylandia_zebra', 'homo_sapiens', 'ENSG00000002745'),\n",
       " ('latimeria_chalumnae', 'homo_sapiens', 'ENSG00000185774'),\n",
       " ('pelodiscus_sinensis', 'homo_sapiens', 'ENSG00000136010'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000076770'),\n",
       " ('petromyzon_marinus', 'homo_sapiens', 'ENSG00000138166')]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_list_of_hidden_genes_with_ortholog_id(sorted_gene_scores: list, ortho_df: pd.DataFrame):\n",
    "    missing_present_ensid_tuples = list()\n",
    "    sorted_gene_ids = sorted_gene_scores.index.tolist()\n",
    "    \n",
    "    for gene in sorted_gene_ids:\n",
    "        species_with_hidden = ortho_df.loc[ortho_df[gene].isna(), gene].index.tolist()\n",
    "        for species in species_with_hidden:\n",
    "            # preferably human ortholog (first in table -> index 0)\n",
    "            species_with_ortholog = ortho_df.loc[ortho_df[gene].notna(), gene].index.tolist()\n",
    "            if len(species_with_ortholog) > 0:\n",
    "                specie_with_ortholog = species_with_ortholog[0]\n",
    "            else:\n",
    "                print(\"Gene is not present in any organism! No orthologs!\")\n",
    "                continue\n",
    "            ensembl_id_ortholog = ortho_df.loc[specie_with_ortholog, gene]\n",
    "\n",
    "        missing_present_ensid_tuples.append((species, specie_with_ortholog, ensembl_id_ortholog))\n",
    "        \n",
    "    return missing_present_ensid_tuples\n",
    "    \n",
    "hidden_genes_tuples = get_list_of_hidden_genes_with_ortholog_id(prioritized_score_df, orthology_df)\n",
    "hidden_genes_tuples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "93e792da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_genes_to_search_to_json(genes_to_search: list, output_file: pathlib.Path):\n",
    "    with open(output_file, \"w\") as f:\n",
    "        json.dump(genes_to_search, f)\n",
    "        \n",
    "save_genes_to_search_to_json(missing_present_ensid_tuples, hidden_genes_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
